File: Task3/Java/Library.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.StringJoiner;

// Перечисление для типов команд (не используется напрямую в processCommand, но сохранено для соответствия)
enum Type {
	CREATE_TRAIN,
	TRAINS_FOR_TOWN,
	TOWNS_FOR_TRAIN,
	TRAINS,
	CREATE_TRAM,
	TRAMS_IN_STOP,
	STOPS_IN_TRAM,
	TRAMS,
	CREATE_PLANE,
	PLANES_FOR_TOWN,
	TOWNS_FOR_PLANE,
	PLANES,
	CREATE_TRL,
	TRL_IN_STOP,
	STOPS_IN_TRL,
	TRLS,
}

// Структура для трамвая (для варианта 2)
class Tram {

	String name;
	List<String> stops;

	public Tram(String name, List<String> stops) {
		this.name = name;
		this.stops = stops;
	}
}

public class Library {

	// Функция для обработки команд (для варианта 2 - трамваи)
	public static void processCommand(
		String command,
		String args,
		List<Tram> trams
	) {
		if (command.equals("CREATE_TRAM")) {
			String[] parts = args.split(" ", 2); // Разделяем имя трамвая и остальное
			String tramName = parts[0];

			String[] stopParts = parts[1].split(" "); // Разделяем количество остановок и сами остановки

			List<String> stops = new ArrayList<>();
			for (int i = 1; i < stopParts.length; i++) { // Начинаем с индекса 1, так как 0 - это количество
				stops.add(stopParts[i]);
			}
			trams.add(new Tram(tramName, stops));
		} else if (command.equals("TRAMS_IN_STOP")) {
			String stopName = args;
			List<String> tramsInThisStop = new ArrayList<>();
			for (Tram tram : trams) {
				for (String stop : tram.stops) {
					if (stop.equals(stopName)) {
						tramsInThisStop.add(tram.name);
						break;
					}
				}
			}
			if (tramsInThisStop.isEmpty()) {
				System.out.println("Trams is absent");
			} else {
				for (String tramName : tramsInThisStop) {
					System.out.println(tramName);
				}
			}
		} else if (command.equals("STOPS_IN_TRAM")) {
			String tramName = args;
			boolean found = false;
			for (Tram tram : trams) {
				if (tram.name.equals(tramName)) {
					found = true;
					Map<String, List<String>> stopsAndTrams = new HashMap<>();
					for (String stop : tram.stops) {
						stopsAndTrams.put(stop, new ArrayList<>()); // Инициализируем список для каждой остановки
						for (Tram otherTram : trams) {
							if (!otherTram.name.equals(tramName)) { // Исключаем текущий трамвай
								for (String otherStop : otherTram.stops) {
									if (otherStop.equals(stop)) {
										stopsAndTrams
											.get(stop)
											.add(otherTram.name);
									}
								}
							}
						}
					}
					for (String stop : tram.stops) {
						System.out.print("Stop " + stop + ": ");
						List<String> tramsForStop = stopsAndTrams.get(stop);
						if (tramsForStop != null && !tramsForStop.isEmpty()) {
							StringJoiner sj = new StringJoiner(" ");
							for (String tName : tramsForStop) {
								sj.add(tName);
							}
							System.out.println(sj.toString());
						} else {
							System.out.println("0");
						}
					}
					break;
				}
			}
			if (!found) {
				System.out.println("Trams is absent");
			}
		} else if (command.equals("TRAMS")) {
			if (trams.isEmpty()) {
				System.out.println("Trams is absent");
			} else {
				for (Tram tram : trams) {
					System.out.print("TRAM " + tram.name + ":");
					for (String stop : tram.stops) {
						System.out.print(" " + stop);
					}
					System.out.println();
				}
			}
		}
	}
}

File: Task3/C++/library.cpp
#include "library.h"

void processCommand(const Type command, const std::string &args, std::vector<Tram> &trams) {
	if (command == Type::CREATE_TRAM) {
		std::stringstream ss(args);
		std::string tramName;
		int numStops;
		ss >> tramName >> numStops;
		if (numStops < 2)
			return;
		for (Tram tram : trams) {
			if (tram.name == tramName)
				return;
		}
		Tram newTram;
		newTram.name = tramName;
		for (int i = 0; i < numStops; ++i) {
			std::string stop;
			ss >> stop;
			newTram.stops.push_back(stop);
		}
		if (newTram.stops.size() == 2) {
			if (newTram.stops.at(0) == newTram.stops.at(1))
				return;
		}
		trams.push_back(newTram);

	} else if (command == Type::TRAMS_IN_STOP) {
		std::string stopName = args;
		std::vector<std::string> tramsInThisStop;
		for (const auto &tram : trams) {
			for (const auto &stop : tram.stops) {
				if (stop == stopName) {
					tramsInThisStop.push_back(tram.name);
					break;
				}
			}
		}
		if (tramsInThisStop.empty()) {
			std::cout << "Trams is absent" << std::endl;
		} else {
			for (const auto &tramName : tramsInThisStop) {
				std::cout << tramName << std::endl;
			}
		}

	} else if (command == Type::STOPS_IN_TRAM) {
		std::string tramName = args;
		bool found = false;
		for (const auto &tram : trams) {
			if (tram.name == tramName) {
				found = true;
				std::map<std::string, std::vector<std::string>> stopsAndTrams;
				for (const auto &stop : tram.stops) {
					for (const auto &otherTram : trams) {
						if (otherTram.name != tramName) { // Исключаем текущий трамвай
							for (const auto &otherStop : otherTram.stops) {
								if (otherStop == stop) {
									stopsAndTrams[stop].push_back(otherTram.name);
								}
							}
						}
					}
				}
				for (const auto &stop : tram.stops) {
					std::cout << "Stop " << stop << ": ";
					if (stopsAndTrams.count(stop) && !stopsAndTrams[stop].empty()) {
						for (size_t i = 0; i < stopsAndTrams[stop].size(); ++i) {
							std::cout << stopsAndTrams[stop][i];
							if (i < stopsAndTrams[stop].size() - 1) {
								std::cout << " ";
							}
						}
						std::cout << std::endl;
					} else {
						std::cout << "0" << std::endl;
					}
				}
				break;
			}
		}
		if (!found) {
			std::cout << "Trams is absent" << std::endl;
		}

	} else if (command == Type::TRAMS) {
		if (trams.empty()) {
			std::cout << "Trams is absent" << std::endl;
		} else {
			for (const auto &tram : trams) {
				std::cout << "TRAM " << tram.name << ":";
				for (const auto &stop : tram.stops) {
					std::cout << " " << stop;
				}
				std::cout << std::endl;
			}
		}
	}
}
